documents/MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,947 @@
+object that contains a description of the PUBLISH packet. The [publish](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt5/Mqtt5Client.html#publish(software.amazon.awssdk.crt.mqtt5.packets.PublishPacket))
+operation takes a `PublishPacket` instance and returns a promise that contains a [PublishResult](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt5/PublishResult.html).
+The returned `PublishResult` will contain different data depending on the `QoS` used in the publish.
+
+* For QoS 0 (AT_MOST_ONCE): Calling `getValue` will return `null` and the promise will be complete as soon as the packet
+has been written to the socket.
+* For QoS 1 (AT_LEAST_ONCE): Calling `getValue` will return a [PubAckPacket](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt5/packets/PubAckPacket.html)
+and the promise will be complete as soon as the PUBACK is received from the broker.
+
+If the operation fails for any reason before these respective completion events, the promise is rejected with a descriptive
+error. You should always check the reason code of a [PubAckPacket](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt5/packets/PubAckPacket.html)
+completion to determine if a QoS 1 publish operation actually succeeded.
+
+#### Example of publishing in the v1 SDK
+
+```java
+// Blocking API.
+client.publish("my/topic", AWSIotQos.QOS0, "hello");
+```
+
+```java
+// Non-blocking API.
+public class MyMessage extends AWSIotMessage {
+    @Override
+    public void onSuccess() {}
+    @Override
+    public void onFailure() {}
+    @Override
+    public void onTimeout() {}
+}
+
+MyMessage message = new MyMessage("my/topic", AWSIotQos.QOS0, "hello");
+long timeout = 3000;  // milliseconds
+client.publish(message, timeout);
+```
+
+#### Example of publishing in the v2 SDK
+
+```java
+PublishPacketBuilder publishBuilder =
+        new PublishPacketBuilder("my/topic", QOS.AT_MOST_ONCE, "hello".getBytes());
+CompletableFuture<PublishResult> published = client.publish(publishBuilder.build());
+PublishResult result = published.get(60, TimeUnit.SECONDS);
+```
+
+### Subscribe
+
+The v1 SDK provides blocking and non-blocking APIs for subscribing. To subscribe to a topic in the v1 SDK, you should
+provide an instance of [AWSIotTopic](http://aws-iot-device-sdk-java-docs.s3-website-us-east-1.amazonaws.com/com/amazonaws/services/iot/client/AWSIotTopic.html)
+to the [subscribe](http://aws-iot-device-sdk-java-docs.s3-website-us-east-1.amazonaws.com/com/amazonaws/services/iot/client/core/AbstractAwsIotClient.html#subscribe-com.amazonaws.services.iot.client.AWSIotTopic-)
+operation. AWSIotTopic object (or, usually, an object of a children class) implements `onMessageReceived` method which
+will be called on receiving a new message. If you try to subscribe to a topic that is not allowed by a policy, AWS IoT
```

Comment:
Same as above for Publish. v2 MQTT5 client doesn't get the connection closed but instead receives a suback with a "Not Authorized" reason code. If we call out this close connection behavior for v1, we should highlight the improved behavior in v2 MQTT5. Doesn't have to go in depth but the difference should either be included or the disconnect behavior description for v1 removed as it serves no purpose.